// @datorama/akita/esm2015/src/internal/global-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class AkitaGlobals {
    constructor() {
        this.skipAction = false;
        this.skipTransactionMsg = false;
        this.currentT = [];
        this.activeTransactions = 0;
    }
    /**
     * @param {?} _action
     * @return {?}
     */
    setAction(_action) {
        if (this.customAction) {
            this.currentAction = this.customAction;
            this.customAction = null;
            this.skipTransactionMsg = false;
        }
        else {
            if (this.activeTransactions === 0) {
                this.currentAction = _action;
            }
        }
        if (this.activeTransactions > 0) {
            this.currentT.push(_action);
        }
    }
    /**
     * @param {?} action
     * @param {?=} skipTransactionMsg
     * @return {?}
     */
    setCustomAction(action, skipTransactionMsg = false) {
        this.currentAction = this.customAction = action;
        this.skipTransactionMsg = skipTransactionMsg;
    }
    /**
     * @param {?=} skip
     * @return {?}
     */
    setSkipAction(skip = true) {
        this.skipAction = skip;
    }
}
function AkitaGlobals_tsickle_Closure_declarations() {
    /** @type {?} */
    AkitaGlobals.prototype.customAction;
    /** @type {?} */
    AkitaGlobals.prototype.currentAction;
    /** @type {?} */
    AkitaGlobals.prototype.skipAction;
    /** @type {?} */
    AkitaGlobals.prototype.skipTransactionMsg;
    /** @type {?} */
    AkitaGlobals.prototype.currentT;
    /** @type {?} */
    AkitaGlobals.prototype.activeTransactions;
    /** @type {?} */
    AkitaGlobals.prototype.batchTransaction;
}
/** @type {?} */
export const __globalState = new AkitaGlobals();
