// @datorama/akita-ngdevtools/bundles/datorama-akita-ngdevtools.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@datorama/akita')) :
    typeof define === 'function' && define.amd ? define('@datorama/akita-ngdevtools', ['exports', '@angular/core', '@datorama/akita'], factory) :
    (factory((global.datorama = global.datorama || {}, global.datorama['akita-ngdevtools'] = {}),global.ng.core,null));
}(this, (function (exports,core,akita) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DEVTOOLS_OPTIONS = new core.InjectionToken('DevtoolsOptions');
    var AkitaDevtools = /** @class */ (function () {
        function AkitaDevtools(ngZone, options) {
            this.ngZone = ngZone;
            this.options = options;
            akita.akitaDevtools(this.ngZone, this.options);
        }
        /** @nocollapse */
        AkitaDevtools.ctorParameters = function () {
            return [
                { type: core.NgZone },
                { type: undefined, decorators: [{ type: core.Inject, args: [DEVTOOLS_OPTIONS,] }] }
            ];
        };
        return AkitaDevtools;
    }());
    /**
     * @return {?}
     */
    function d() { }
    /**
     * @param {?} akitaDevtools
     * @return {?}
     */
    function init(akitaDevtools) {
        return d;
    }
    var AkitaNgDevtools = /** @class */ (function () {
        function AkitaNgDevtools() {
        }
        /**
         * @param {?=} options
         * @return {?}
         */
        AkitaNgDevtools.forRoot = /**
         * @param {?=} options
         * @return {?}
         */
            function (options) {
                return {
                    ngModule: AkitaNgDevtools,
                    providers: [
                        AkitaDevtools,
                        {
                            provide: DEVTOOLS_OPTIONS,
                            useValue: options
                        },
                        {
                            provide: core.APP_INITIALIZER,
                            useFactory: init,
                            deps: [AkitaDevtools],
                            multi: true
                        }
                    ]
                };
            };
        AkitaNgDevtools.decorators = [
            { type: core.NgModule, args: [{},] },
        ];
        return AkitaNgDevtools;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.d = d;
    exports.init = init;
    exports.DEVTOOLS_OPTIONS = DEVTOOLS_OPTIONS;
    exports.AkitaDevtools = AkitaDevtools;
    exports.AkitaNgDevtools = AkitaNgDevtools;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
