// @datorama/akita/esm5/src/api/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * An interface to represent objects
 *
 * \@example
 * const dashboards: HashMap<Dashboard> = {
 *   1: Dashboard
 * }
 * @record
 * @template T
 */
export function HashMap() { }
function HashMap_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [id: string]: T;
    */
}
/**
 *  The Entity State is a predefined generic interface for a given collection
 *  with the following interface:
 *
 *  `ids`: An array of all the primary ids in the collection
 *  `entities`: A dictionary of entities in the collection indexed by the primary id
 * @record
 * @template E, ErrorT
 */
export function EntityState() { }
function EntityState_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    EntityState.prototype.entities;
    /** @type {?|undefined} */
    EntityState.prototype.ids;
    /** @type {?|undefined} */
    EntityState.prototype.loading;
    /** @type {?|undefined} */
    EntityState.prototype.error;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * @record
 * @template E
 */
export function Entities() { }
function Entities_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    Entities.prototype.entities;
    /** @type {?|undefined} */
    Entities.prototype.ids;
}
/**
 * Interface for stores that needs an active indication
 * @record
 * @template T
 */
export function ActiveState() { }
function ActiveState_tsickle_Closure_declarations() {
    /** @type {?} */
    ActiveState.prototype.active;
}
