// @datorama/akita/esm5/src/internal/global-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AkitaGlobals = /** @class */ (function () {
    function AkitaGlobals() {
        this.skipAction = false;
        this.skipTransactionMsg = false;
        this.currentT = [];
        this.activeTransactions = 0;
    }
    /**
     * @param {?} _action
     * @return {?}
     */
    AkitaGlobals.prototype.setAction = /**
     * @param {?} _action
     * @return {?}
     */
    function (_action) {
        if (this.customAction) {
            this.currentAction = this.customAction;
            this.customAction = null;
            this.skipTransactionMsg = false;
        }
        else {
            if (this.activeTransactions === 0) {
                this.currentAction = _action;
            }
        }
        if (this.activeTransactions > 0) {
            this.currentT.push(_action);
        }
    };
    /**
     * @param {?} action
     * @param {?=} skipTransactionMsg
     * @return {?}
     */
    AkitaGlobals.prototype.setCustomAction = /**
     * @param {?} action
     * @param {?=} skipTransactionMsg
     * @return {?}
     */
    function (action, skipTransactionMsg) {
        if (skipTransactionMsg === void 0) { skipTransactionMsg = false; }
        this.currentAction = this.customAction = action;
        this.skipTransactionMsg = skipTransactionMsg;
    };
    /**
     * @param {?=} skip
     * @return {?}
     */
    AkitaGlobals.prototype.setSkipAction = /**
     * @param {?=} skip
     * @return {?}
     */
    function (skip) {
        if (skip === void 0) { skip = true; }
        this.skipAction = skip;
    };
    return AkitaGlobals;
}());
export { AkitaGlobals };
function AkitaGlobals_tsickle_Closure_declarations() {
    /** @type {?} */
    AkitaGlobals.prototype.customAction;
    /** @type {?} */
    AkitaGlobals.prototype.currentAction;
    /** @type {?} */
    AkitaGlobals.prototype.skipAction;
    /** @type {?} */
    AkitaGlobals.prototype.skipTransactionMsg;
    /** @type {?} */
    AkitaGlobals.prototype.currentT;
    /** @type {?} */
    AkitaGlobals.prototype.activeTransactions;
    /** @type {?} */
    AkitaGlobals.prototype.batchTransaction;
}
/** @type {?} */
export var __globalState = new AkitaGlobals();
