// @datorama/akita/esm2015/src/plugins/dirty-check/dirty-check-plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { AkitaPlugin } from '../plugin';
import { QueryEntity } from '../../api/query-entity';
import { BehaviorSubject, merge } from 'rxjs';
import { distinctUntilChanged, map, skip } from 'rxjs/operators';
import { coerceArray, isFunction, isUndefined, toBoolean } from '../../internal/utils';
import { __globalState } from '../../internal/global-state';
/** @type {?} */
export const dirtyCheckDefaultParams = {
    comparator: (head, current) => JSON.stringify(head) !== JSON.stringify(current)
};
/**
 * @template Entity, StoreState
 */
export class DirtyCheckPlugin extends AkitaPlugin {
    /**
     * @param {?} query
     * @param {?=} params
     * @param {?=} _entityId
     */
    constructor(query, params, _entityId) {
        super(query);
        this.query = query;
        this.params = params;
        this._entityId = _entityId;
        this.dirty = new BehaviorSubject(false);
        this.active = false;
        this.isDirty$ = this.dirty.asObservable().pipe(distinctUntilChanged());
        this.params = Object.assign({}, dirtyCheckDefaultParams, params);
        if (this.params.watchProperty) {
            /** @type {?} */
            let watchProp = this.params.watchProperty;
            watchProp = coerceArray(watchProp);
            if (watchProp.includes('entities') && !watchProp.includes('ids') && query instanceof QueryEntity) {
                watchProp.push('ids');
            }
            this.params.watchProperty = watchProp;
        }
    }
    /**
     * @return {?}
     */
    getHead() {
        return this.head;
    }
    /**
     * @return {?}
     */
    activate() {
        this.head = this._getHead();
        /** *
         * if we are tracking specific properties select only the relevant ones
          @type {?} */
        const source = this.params.watchProperty
            ? (/** @type {?} */ (this.params.watchProperty)).map(prop => this.query.select(state => state[prop]).pipe(map(val => ({ val, __akitaKey: prop }))))
            : [this.selectSource(this._entityId)];
        this.subscription = merge(...source)
            .pipe(skip(1))
            .subscribe((currentState) => {
            if (isUndefined(this.head))
                return;
            /** *
             * __akitaKey is used to determine if we are tracking a specific property or a store change
              @type {?} */
            const head = currentState.__akitaKey ? this.head[/** @type {?} */ (currentState.__akitaKey)] : this.head;
            /** @type {?} */
            const compareTo = currentState.__akitaKey ? currentState.val : currentState;
            /** @type {?} */
            const isChange = this.params.comparator(head, compareTo);
            this.updateDirtiness(isChange);
        });
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    reset(params = {}) {
        /** @type {?} */
        let currentValue = this.head;
        if (isFunction(params.updateFn)) {
            if (this.isEntityBased(this._entityId)) {
                currentValue = params.updateFn(this.head, (/** @type {?} */ (this.getQuery())).getEntity(this._entityId));
            }
            else {
                currentValue = params.updateFn(this.head, (/** @type {?} */ (this.getQuery())).getSnapshot());
            }
        }
        /** *
         * If we are watching specific props compare them, if not compare the entire store
          @type {?} */
        const update = this.params.watchProperty ? this.compareProp(currentValue) : this._getHead() !== currentValue;
        if (update) {
            __globalState.setCustomAction({ type: `@DirtyCheck - Revert` });
            this.updateStore(currentValue, this._entityId);
        }
    }
    /**
     * @return {?}
     */
    setHead() {
        if (!this.active) {
            this.activate();
            this.active = true;
        }
        else {
            this.head = this._getHead();
        }
        this.updateDirtiness(false);
        return this;
    }
    /**
     * @return {?}
     */
    isDirty() {
        return toBoolean(this.dirty.value);
    }
    /**
     * @return {?}
     */
    hasHead() {
        return toBoolean(this.getHead());
    }
    /**
     * @return {?}
     */
    destroy() {
        this.head = null;
        this.subscription && this.subscription.unsubscribe();
    }
    /**
     * @param {?} isDirty
     * @return {?}
     */
    updateDirtiness(isDirty) {
        this.dirty.next(isDirty);
    }
    /**
     * @return {?}
     */
    _getHead() {
        /** @type {?} */
        let head = /** @type {?} */ (this.getSource(this._entityId));
        if (this.params.watchProperty) {
            head = (/** @type {?} */ (this.params.watchProperty)).reduce((_head, prop) => {
                _head[prop] = (/** @type {?} */ (head))[prop];
                return _head;
            }, /** @type {?} */ ({}));
        }
        return head;
    }
    /**
     * @param {?} currentState
     * @return {?}
     */
    compareProp(currentState) {
        /** @type {?} */
        const propKeys = Object.keys(currentState);
        /** @type {?} */
        const head = this._getHead();
        return propKeys.some(propKey => currentState[propKey] !== head[propKey]);
    }
}
function DirtyCheckPlugin_tsickle_Closure_declarations() {
    /** @type {?} */
    DirtyCheckPlugin.prototype.head;
    /** @type {?} */
    DirtyCheckPlugin.prototype.dirty;
    /** @type {?} */
    DirtyCheckPlugin.prototype.subscription;
    /** @type {?} */
    DirtyCheckPlugin.prototype.active;
    /** @type {?} */
    DirtyCheckPlugin.prototype.isDirty$;
    /** @type {?} */
    DirtyCheckPlugin.prototype.query;
    /** @type {?} */
    DirtyCheckPlugin.prototype.params;
    /** @type {?} */
    DirtyCheckPlugin.prototype._entityId;
}
