// @datorama/akita/esm5/src/plugins/persist-form/persist-ng-form-plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AkitaPlugin } from '../plugin';
import { debounceTime } from 'rxjs/operators';
import { __globalState } from '../../internal/global-state';
import { getValue, isString, setValue } from '../../internal/utils';
/**
 * @template T
 */
var /**
 * @template T
 */
PersistNgFormPlugin = /** @class */ (function (_super) {
    tslib_1.__extends(PersistNgFormPlugin, _super);
    function PersistNgFormPlugin(query, factoryFnOrPath, params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this, query) || this;
        _this.query = query;
        _this.factoryFnOrPath = factoryFnOrPath;
        _this.params = params;
        _this.params = tslib_1.__assign({ debounceTime: 300, formKey: 'akitaForm', emitEvent: false }, params);
        _this.isKeyBased = isString(factoryFnOrPath);
        return _this;
    }
    /**
     * @param {?} form
     * @return {?}
     */
    PersistNgFormPlugin.prototype.setForm = /**
     * @param {?} form
     * @return {?}
     */
    function (form) {
        this.form = form;
        this.activate();
        return this;
    };
    /**
     * @param {?=} initialState
     * @return {?}
     */
    PersistNgFormPlugin.prototype.reset = /**
     * @param {?=} initialState
     * @return {?}
     */
    function (initialState) {
        var _a;
        /** @type {?} */
        var value;
        if (initialState) {
            value = initialState;
        }
        else {
            value = this.isKeyBased ? this.initialValue : (/** @type {?} */ (this)).factoryFnOrPath();
        }
        this.form.patchValue(value);
        /** @type {?} */
        var storeValue = this.isKeyBased ? setValue(this.getStore()._value(), this.getStore().storeName + "." + this.factoryFnOrPath, value) : (_a = {}, _a[this.params.formKey] = value, _a);
        this.updateStore(storeValue);
    };
    /**
     * @return {?}
     */
    PersistNgFormPlugin.prototype.activate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var _a;
        /** @type {?} */
        var path;
        if (this.isKeyBased) {
            path = this.getStore().storeName + "." + this.factoryFnOrPath;
            this.initialValue = getValue(this.getStore()._value(), path);
            this.form.patchValue(this.initialValue, { emitEvent: this.params.emitEvent });
        }
        else {
            if (!(/** @type {?} */ (this.getQuery().getSnapshot()))[this.params.formKey]) {
                __globalState.setAction({ type: '@PersistNgFormPlugin activate' });
                this.updateStore((_a = {}, _a[this.params.formKey] = (/** @type {?} */ (this)).factoryFnOrPath(), _a));
            }
            this.query.selectOnce(function (state) { return (/** @type {?} */ (state))[_this.params.formKey]; }).subscribe(function (formValue) { return _this.form.patchValue(formValue); });
        }
        this.formChanges = this.form.valueChanges.pipe(debounceTime(this.params.debounceTime)).subscribe(function (value) {
            __globalState.setAction({ type: '@PersistForm - Update' });
            /** @type {?} */
            var newState;
            if (_this.isKeyBased) {
                newState = function (state) { return setValue(state, path, value); };
            }
            else {
                newState = function () {
                    var _a;
                    return (_a = {}, _a[_this.params.formKey] = value, _a);
                };
            }
            _this.updateStore(newState(_this.getStore()._value()));
        });
    };
    /**
     * @return {?}
     */
    PersistNgFormPlugin.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.formChanges && this.formChanges.unsubscribe();
    };
    return PersistNgFormPlugin;
}(AkitaPlugin));
/**
 * @template T
 */
export { PersistNgFormPlugin };
function PersistNgFormPlugin_tsickle_Closure_declarations() {
    /** @type {?} */
    PersistNgFormPlugin.prototype.formChanges;
    /** @type {?} */
    PersistNgFormPlugin.prototype.form;
    /** @type {?} */
    PersistNgFormPlugin.prototype.isKeyBased;
    /** @type {?} */
    PersistNgFormPlugin.prototype.initialValue;
    /** @type {?} */
    PersistNgFormPlugin.prototype.query;
    /** @type {?} */
    PersistNgFormPlugin.prototype.factoryFnOrPath;
    /** @type {?} */
    PersistNgFormPlugin.prototype.params;
}
